lib/ecto/query/builder/from.ex

```diff
@@ -80,12 +80,13 @@ defmodule Ecto.Query.Builder.From do
       )
     end
 
-    case prefix do
-      nil -> :ok
-      {:ok, prefix} when is_binary(prefix) or is_nil(prefix) -> :ok
-      _ -> Builder.error!("`prefix` must be a compile time string, got: `#{Macro.to_string(prefix)}`")
+    prefix = case prefix do
+      nil -> nil
+      {:ok, prefix} when is_binary(prefix) or is_nil(prefix) -> {:ok, prefix}
+      {:ok, {:^, _, [prefix]}} -> {:ok, prefix}
```

Comment:
```suggestion
      {:ok, {:^, _, [prefix]}} -> {:ok, quote(do: Ecto.Query.Builder.From.prefix!(unquote(prefix))}
```

The idea is that we will wrap it in a code that will validate it at runtime. :) The same needs to be done for join!

Reply:
You mean like this? https://github.com/elixir-ecto/ecto/pull/4098/commits/6e546ac962c34726afb23d743d090a3add3b25ac

Reply:
We were unsure whether to use `Builder.error!` at this point as it's not really an `Ecto.Query.CompileError`?